File: C++/Oct22-3.cpp
#include <bits/stdc++.h>
using namespace std;


void permute(string str, string out)
{
    if (str.size() == 0)
    {
        cout << out << endl;
        return;
    }

    for (int i = 0; i < str.size(); i++)
    {
        permute(str.substr(1), out + str[0]);
 
        rotate(str.begin(), str.begin() + 1, str.end());
    }
}


int main()
{
    string str = "AGSF";
    permute(str, "");
    return 0;
}

File: JavaScript/Oct22-3.js
function *permute(a, n = a.length) {
  if (n <= 1) yield a.slice();
  else for (let i = 0; i < n; i++) {
    yield *permute(a, n - 1);
    const j = n % 2 ? 0 : i;
    [a[n-1], a[j]] = [a[j], a[n-1]];
  }
}

console.log(Array.from(permute("AGSF".split('')))
.map(perm => perm.join(''))
.filter((el, idx, self) => (self.indexOf(el) === idx)));